spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/sql/GcpCloudSqlAutoConfiguration.java

```diff
@@ -19,144 +19,233 @@
 import org.springframework.boot.autoconfigure.jdbc.DataSourceProperties;
+import org.springframework.boot.autoconfigure.jdbc.JndiDataSourceAutoConfiguration;
+import org.springframework.boot.autoconfigure.jdbc.XADataSourceAutoConfiguration;
 import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.cloud.gcp.autoconfigure.core.AppEngineCondition;
 import org.springframework.cloud.gcp.autoconfigure.core.GcpContextAutoConfiguration;
 import org.springframework.cloud.gcp.autoconfigure.core.GcpProperties;
 import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Conditional;
 import org.springframework.context.annotation.Configuration;
 import org.springframework.context.annotation.Import;
 import org.springframework.context.annotation.Primary;
+import org.springframework.jdbc.datasource.embedded.EmbeddedDatabaseType;
 import org.springframework.util.StringUtils;
 
 /**
  * Provides Google Cloud SQL instance connectivity through Spring JDBC by providing only a
  * database and instance connection name.
  *
  * @author João André Martins
+ * @author Artem Bilan
  */
 @Configuration
+@ConditionalOnClass({ DataSource.class, EmbeddedDatabaseType.class })
+@ConditionalOnProperty(name = "spring.cloud.gcp.sql.enabled", havingValue = "true", matchIfMissing = true)
 @EnableConfigurationProperties({ GcpCloudSqlProperties.class, DataSourceProperties.class })
-@ConditionalOnProperty(
-		name = "spring.cloud.gcp.sql.enabled", havingValue = "true", matchIfMissing = true)
-@AutoConfigureBefore(DataSourceAutoConfiguration.class)
-@AutoConfigureAfter({GcpContextAutoConfiguration.class,
-		AppEngineJdbcInfoProviderAutoConfiguration.class,
-		MySqlJdbcInfoProviderAutoConfiguration.class,
-		PostgreSqlJdbcInfoProviderAutoConfiguration.class})
-@Import({AppEngineJdbcInfoProviderAutoConfiguration.class,
-		MySqlJdbcInfoProviderAutoConfiguration.class,
-		PostgreSqlJdbcInfoProviderAutoConfiguration.class})
+@AutoConfigureBefore({
+		DataSourceAutoConfiguration.class,
+		JndiDataSourceAutoConfiguration.class,
+		XADataSourceAutoConfiguration.class })
+@AutoConfigureAfter(GcpContextAutoConfiguration.class)
 public class GcpCloudSqlAutoConfiguration {
 
 	public final static String INSTANCE_CONNECTION_NAME_HELP_URL =
 			"https://github.com/spring-cloud/spring-cloud-gcp/tree/master/"
-			+ "spring-cloud-gcp-starters/spring-cloud-gcp-starter-sql"
-			+ "#google-cloud-sql-instance-connection-name";
+					+ "spring-cloud-gcp-starters/spring-cloud-gcp-starter-sql"
+					+ "#google-cloud-sql-instance-connection-name";
 
 	private static final Log LOGGER = LogFactory.getLog(GcpCloudSqlAutoConfiguration.class);
 
-	private final GcpCloudSqlProperties gcpCloudSqlProperties;
 
-	private final GcpProperties gcpProperties;
+	/**
+	 * The AppEngine Configuration for the {@link AppEngineCloudSqlJdbcInfoProvider}.
+	 */
+	@ConditionalOnClass(com.google.cloud.sql.mysql.SocketFactory.class)
```

Comment:
@snicoll mentioned using `@ConditionalOnClass` in methods to be an anti-pattern, which is why we created all the other autoconfig classes.

Reply:
???
This is not method.
Let me repeat it again with a simplified code:
```
@ConditionalOnClass(com.google.cloud.sql.mysql.SocketFactory.class)
static class AppEngineJdbcInfoProviderConfiguration {
```
Where do you see method here, please?

Reply:
Sorry, didn't realise this was a class, my bad!